dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -0,0 +1,358 @@
+
+        THROW_HR_IF(E_UNEXPECTED, notificationId == 0);
+
+        winrt::Microsoft::Windows::AppNotifications::implementation::AppNotification* notificationImpl = get_self< winrt::Microsoft::Windows::AppNotifications::implementation::AppNotification>(notification);
+        notificationImpl->SetNotificationId(notificationId);
+    }
+
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Microsoft::Windows::AppNotifications::AppNotificationProgressResult> AppNotificationManager::UpdateAsync(winrt::Microsoft::Windows::AppNotifications::AppNotificationProgressData const data, hstring const tag, hstring const group)
+    {
+        THROW_HR_IF_MSG(E_INVALIDARG, tag == winrt::hstring(L""), "Update operation isn't guaranteed to find a specific notification to replace correctly.");
+        THROW_HR_IF_MSG(E_INVALIDARG, data.SequenceNumber() == 0, "Sequence Number for Updates should be greater than 0!");
+
+        std::wstring appId{ RetrieveNotificationAppId() };
+
+        winrt::com_ptr<ToastABI::IToastProgressData> toastProgressData{ winrt::make_self<NotificationProgressData>(data) };
+
+        auto hr = ToastNotifications_UpdateNotificationData(appId.c_str(), tag.c_str(), group.c_str(), toastProgressData.get());
+
+        if (SUCCEEDED(hr))
+        {
+            co_return winrt::AppNotificationProgressResult::Succeeded;
+        }
+        else if (hr == E_NOT_SET)
+        {
+            co_return winrt::AppNotificationProgressResult::AppNotificationNotFound;
+        }
+        else
+        {
+            THROW_HR(hr);
+        }
+    }
+
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Microsoft::Windows::AppNotifications::AppNotificationProgressResult> AppNotificationManager::UpdateAsync(winrt::Microsoft::Windows::AppNotifications::AppNotificationProgressData const data, hstring const tag)
+    {
+        co_return co_await UpdateAsync(data, tag, L"");
+    }
+
+    winrt::Microsoft::Windows::AppNotifications::AppNotificationSetting AppNotificationManager::Setting()
+    {
+        std::wstring appId{ RetrieveNotificationAppId() };
+        DWORD appNotificationSetting{ 0 };
+        ToastNotifications_QuerySettings(appId.c_str(), &appNotificationSetting);
+        return static_cast<winrt::Microsoft::Windows::AppNotifications::AppNotificationSetting>(appNotificationSetting);
+    }
+
+    winrt::Windows::Foundation::IAsyncAction AppNotificationManager::RemoveByIdAsync(uint32_t notificationId)
+    {
+        THROW_HR_IF(E_INVALIDARG, notificationId == 0);
+
+        co_await winrt::resume_background();
+
+        std::wstring appId{ RetrieveNotificationAppId() };
+        THROW_IF_FAILED(ToastNotifications_RemoveToast(appId.c_str(), notificationId));
+    }
+
+    winrt::Windows::Foundation::IAsyncAction AppNotificationManager::RemoveByTagAsync(hstring const tag)
+    {
+        THROW_HR_IF(E_INVALIDARG, tag == winrt::hstring(L""));
+		
+        std::wstring _tag{ tag.c_str() };
```

Comment:
>std::wstring _tag{ tag.c_str() }; [](http://example.com/codeflow?start=8&length=33)

No need for this - you can say `.c_str()` on an `hstring` just fine.

Reply:
@loneursid